### PR TITLE
feat: enhanced type exception handling

### LINKED ISSUE #754: Type Exception Handling for the Mail Helper
#### Issue Summary

The appropriate types necessary for a successful `mail/send` API call is documented [here](https://sendgrid.com/docs/API_Reference/api_v3.html). We want to throw a type exception when the wrong type is provided.

#### Acceptance Criteria

* All parameters for the `mail\send` endpoint are type checked and if they fail the type check and exception is thrown with instructions on how to fix

### PR DESCRIPTION
**Pull request by misantron**
Closes #754 
Closes https://github.com/sendgrid/sendgrid-php/issues/987

### Checklist
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guide] and my PR follows them.
- [x] I updated my branch with the master branch.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added in line documentation to the code I modified

### Short description of what this PR does:
- New `Assert` class for strict type validation (100% code coverage with tests)
- Refactoring using params type assertion

### CODE REVIEW

**Comment by SendGridDX** [comment]
[CLA assistant check](https://cla.sendgrid.com/sendgrid/sendgrid-php?pullRequest=762) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.sendgrid.com/sendgrid/sendgrid-php?pullRequest=762) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.sendgrid.com/check/sendgrid/sendgrid-php?pullRequest=762) it.</sub>

File: lib/helper/Assert.php [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+     * @param string $property
+     * @param string|null $message
+     *
+     * @throws TypeException
+     */
+    public static function email($value, $property, $message = null)
+    {
+        static::string($value, $property);
+
+        if (!\filter_var($value, \FILTER_VALIDATE_EMAIL)) {
```
Comment (peter279k) [inline lib/helper/Assert.php]: Is this `\` really necessary before `FILTER_VALIDATE_EMAIL`  const value?

The `FILTER_VALIDATE_EMAIL` is the const value and I think we don't need to add the `\` before the const values.

Comment (misantron) [reply lib/helper/Assert.php]: It's all about compile time optimization. But I will remove it.

**Comment by misantron** [comment]
Made some optional improvement of TravisCI deployment:
- Add Composer cache to prevent everytime same dependencies installation
- Add Prism mock server run from Docker container instead of download and run from .sh script (https://docs.stoplight.io/testing/continuous-integration/travis)

**Comment by peter279k** [comment]
@misantron, I don't think the `.travis.yml` enhancement is related to this PR.

Please split two PRs.

**Comment by misantron** [comment]
@peter279k These builds throw an error because of Prism server:
https://travis-ci.org/sendgrid/sendgrid-php/builds/447213062
https://travis-ci.org/sendgrid/sendgrid-php/builds/447216145
This changes is an attempt to solve the problem.

**Comment by peter279k** [comment]
All right. And you can create new PR to fix that. It's out of range for this PR.
You can write comment in this PR to mention this issue and refer another PR.

**Comment by misantron** [comment]
Ok. Should I rollback the changes?

**Comment by misantron** [comment]
#766 @peter279k PR for TravisCI is ready

**Comment by thinkingserious** [comment]
Hello @misantron,<br /><br/>Thanks again for the PR!<br />                                              <br />It's <a href="https://sendgrid.com/blog/hacktoberfest-2018-has-arrived/">HACKTOBERFEST</a>! We want to show our appreciation by sending                                              you some special Hacktoberfest swag. If you have not already, could you please fill out                                              <a href="https://dx.sendgrid.com/hacktoberfest/join">this form</a> so we can send it to you? Thanks!                                             <br /><br />Team SendGrid DX

**Comment by kampalex** [comment]
This is not a official review, just some thoughts about contribution yet:
**Good work!**

In file `lib/mail/Asm.php`, I suggest to replace 'Twilio' in the `TypeException` message because it's unnecessary:
`...must be an instance of Twilio SendGrid\Mail\...`
to
`...must be an instance of SendGrid\Mail\...`

The `Assert::email` method doesn't contain the invalid 'string value' in the `TypeException` message (as requested in #987). Currently, it only contains the property name.

The `EmailAddress::isValidEmailAddress` logic can link to the `Assert::email` method like this:
```php
try {
  Assert::email($emailAddress, 'emailAddress');
  return true;
} catch (TypeException $exception) {
  return false;
}
```
But as seen in all current usages (for now, only `Mail::setFrom` is left), this method might be unnecessary because every 'else' scenario leads to `TypeException`. Despite it's a public static method, the intent usage is 'this library only'.

The `Mail::setFrom` now allows an invalid string; I suggest to change it to:
```php
  if ($email instanceof From) {
    $this->from = $email;
  } else {
    Assert::email($email, 'email', 'Value "$%s" is not an email address.');
    $this->from = new From($email, $name);
  }
```

After digging a bit more, I'm a bit worried because this PR has also breaking changes:
Some methods now expecting explicit class instances in favor of `Assert::isInstanceOf` which can lead to different exception/fatal error.
In PHP 5.6: `PHP Catchable fatal error`
In PHP 7.x: `Fatal error: Uncaught TypeError: Argument 1 passed to ...`

That's it for now!

File: lib/mail/Personalization.php [hunk 2]
```diff
@@ -107,12 +109,8 @@ public function getBccs()
     public function setSubject($subject)
     {
         if (!($subject instanceof Subject)) {
-            //  If not a string, Subject can't be created
-            if (!is_string($subject)) {
-                throw new TypeException(
-                    '$subject must be an instance of SendGrid\Mail\Subject or of type string.'
-                );
-            }
+            Assert::string($subject, 'subject');
```
Comment (childish-sambino) [inline lib/mail/Personalization.php:112]: Messaging needed about being `SendGrid\Mail\Subject or string`.

File: lib/mail/Substitution.php [hunk 3]
```diff
@@ -72,13 +72,17 @@ public function getKey()
...
-        if (!is_string($value) && !is_array($value) && !is_object($value) && !is_bool($value) && !is_numeric($value)) {
-            throw new TypeException("The value '$value' must be of type string, array, object, bool, int, or float.");
-        }
+        Assert::accept($value, 'value', static function ($val) {
+            return \is_string($val)
+                || filter_var($val, FILTER_VALIDATE_INT) !== false
+                || \is_bool($val)
+                || \is_array($val)
+                || \is_object($val);
+        }, '$value must be an array, object, boolean, string or integer.');
```
Comment (childish-sambino) [inline lib/mail/Substitution.php:84]: Floats are also allowed.

File: test/unit/GroupsToDisplayTest.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public function testAddGroupToDisplayWithAlready()
+    {
+        $this->expectException(TypeException::class);
+        $this->expectExceptionMessage('Number of elements in "$groups_to_display" can not be more than 25.');
+
+        $data = range(1, 25);
+        $groups = new GroupsToDisplay($data);
+        $groups->addGroupToDisplay(1);
```
Comment (childish-sambino) [inline test/unit/GroupsToDisplayTest.php:39]: nit: it's possible the last line is not run (which is where the expected exception should come from) since the exception _could_ be thrown during construction.

File: lib/mail/Mail.php [hunk 5]
```diff
@@ -709,7 +709,7 @@ public function addDynamicTemplateDatas(
         $personalizationIndex = null,
         $personalization = null
     ) {
-        $this->addSubstitutions($datas, $personalizationIndex, $personalization);
+        $this->addSubstitutions($datas);
```
Comment (childish-sambino) [inline lib/mail/Mail.php:712]: Revert this.

File: lib/mail/Mail.php [hunk 6]
```diff
@@ -934,16 +934,16 @@ public function getSendAt($personalizationIndex = 0)
...
      */
     public function setFrom($email, $name = null)
     {
         if ($email instanceof From) {
             $this->from = $email;
-        } elseif (EmailAddress::isValidEmailAddress($email)) {
-            $this->from = new From($email, $name);
         } else {
-            throw new TypeException('$email must be valid and of type string.');
+            Assert::email($email, 'email');
```
Comment (childish-sambino) [inline lib/mail/Mail.php:945]: Needs wording about being a `From` as well.

File: lib/mail/Mail.php [hunk 7]
```diff
@@ -1530,15 +1536,9 @@ public function getIpPoolName()
      *                                    mail settings that you can
      *                                    use to specify how you would
      *                                    like this email to be handled
-     * @throws TypeException
      */
-    public function setMailSettings($mail_settings)
+    public function setMailSettings(MailSettings $mail_settings)
```
Comment (childish-sambino) [inline lib/mail/Mail.php:1540]: Compatibility issue.

File: lib/mail/Mail.php [hunk 8]
```diff
@@ -1672,15 +1684,9 @@ public function setSpamCheck($enable = null, $threshold = null, $post_to_url = n
      *                                            would like to track the metrics
      *                                            of how your recipients interact
      *                                            with your email
-     * @throws TypeException
      */
-    public function setTrackingSettings($tracking_settings)
+    public function setTrackingSettings(TrackingSettings $tracking_settings)
```
Comment (childish-sambino) [inline lib/mail/Mail.php:1688]: Compatibility issue.

File: test/phpunit.xml.dist [hunk 9]
```diff
@@ -5,4 +5,9 @@
             <directory>.</directory>
         </testsuite>
     </testsuites>
+    <filter>
```
Comment (childish-sambino) [inline test/phpunit.xml.dist:8]: Why is this needed?

File: lib/helper/Assert.php [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+     *
+     * @param mixed $value
+     * @param string $property
+     * @param string|null $message
+     *
+     * @throws TypeException
+     */
+    public static function email($value, $property, $message = null)
+    {
+        static::string($value, $property);
```
Comment (childish-sambino) [inline lib/helper/Assert.php:48]: Should pass `$message`.

File: lib/helper/Assert.php [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+     * @param mixed $value
+     * @param string $property
+     * @param callable $callback
+     * @param string|null $message
+     *
+     * @throws TypeException
+     */
+    public static function accept($value, $property, $callback, $message = null)
+    {
+        static::isCallable($callback, $property);
```
Comment (childish-sambino) [inline lib/helper/Assert.php:189]: Should be `static::isCallable($callback, 'callback');`

File: test/phpunit.xml.dist (hunk 9 above)
Comment (misantron) [reply test/phpunit.xml.dist:8]: No need to calculate code coverage for not *.php files (README.md etc)

File: test/unit/GroupsToDisplayTest.php (hunk 4 above)
Comment (misantron) [reply test/unit/GroupsToDisplayTest.php:31]: This case test `addGroupToDisplay()`
we add exeactly 25 groups on object construction - ok
than add one more group - got an exception
anything wrong?

**Comment by misantron** [comment]
build on php 7.4 failed because of Prism server error

File: test/unit/GroupsToDisplayTest.php (hunk 4 above)
Comment (childish-sambino) [reply test/unit/GroupsToDisplayTest.php:31]: If the max allowed was incorrectly defined as 24, then this test would still pass but the last line would not get executed because the constructor would be the one throwing the exception.
